File: DataStructures/Tries/Tree.cpp
#include "Tree.h"
Tree::Tree() {
	this->root = nullptr;
}
Tree::~Tree() {
	if (this->root != nullptr) {
		delete this->root;
	}
	this->root = 0;
}

void Tree::Insert(const char word[],int count) {
	if (this->root == nullptr) {
		this->root = new Node();
	}
	this->root->Insert(word,count,0);
}
void Tree::Insert(std::string word) {
	if (this->root == nullptr) {
		this->root = new Node();
	}
	this->root->Insert(word.c_str(), word.length(), 0);
}
bool Tree::Contains(const char word[],int count) {
	if (this->root == nullptr) return false;
	return this->root->Contains(word,count,0);
}
bool Tree::Contains(std::string word) {
	if (this->root == nullptr) return false;
	return this->root->Contains(word.c_str(), word.length(), 0);
}

File: algorithm/_interface/tree.py
import abc


class Tree(metaclass=abc.ABCMeta):
    @abc.abstractmethod
    def insert(self, value) -> None:
        """
        Inserts the value into the tree
        :param value: The value to add to the tree
        :return: None
        """
        pass

    @abc.abstractmethod
    def exists(self, value) -> bool:
        """
        Check if tree contains value
        :param value: value to look for
        :return: Does value exist in tree?
        """
        pass


class Node(metaclass=abc.ABCMeta):
    def __init__(self, value):
        self.value = value

    @abc.abstractmethod
    def insert(self, value) -> None:
        pass

